core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -413,57 +410,76 @@ object ZLayer {
    */
   private trait MemoMap { self =>
 
-    /**
-     * Retrieves a dependency from the memo map if it exists.
-     */
-    def get[E, A, B <: Has[_]](layer: ZLayer[A, E, B]): UIO[Option[B]]
-
-    /**
-     * Stores a dependency in the memo map.
-     */
-    def memoize[E, A, B <: Has[_]](layer: ZLayer[A, E, B], b: B): UIO[Unit]
-
     /**
      * Checks the memo map to see if a dependency exists. If it is, immediately
      * returns it. Otherwise, obtains the dependency, stores it in the memo map,
      * and adds a finalizer to the outer `Managed`.
      */
-    final def getOrElseMemoize[E, A, B <: Has[_]](
+    def getOrElseMemoize[E, A, B <: Has[_]](
       layer: ZLayer[A, E, B],
       finalizerRef: Ref[Exit[Any, Any] => ZIO[Any, Nothing, Any]]
-    ): ZManaged[A, E, B] =
-      ZManaged {
-        get(layer).flatMap {
-          case Some(b) => ZIO.succeedNow(Reservation(ZIO.succeedNow(b), _ => UIO.unit))
-          case None =>
-            layer.scope.flatMap(_.apply(self)).reserve.map { reservation =>
-              Reservation(
-                reservation.acquire.tap(b => memoize(layer, b)),
-                _ =>
-                  ZIO.accessM[A] { a =>
-                    finalizerRef.update { finalizer => exit =>
-                      finalizer(exit) *> reservation.release(exit).provide(a)
-                    }
-                  }
-              )
-            }
-        }
-      }
+    ): ZManaged[A, E, B]
   }
 
   private object MemoMap {
 
     /**
-     * Constructs an empty memo map backed by a `Ref`.
+     * Constructs an empty memo map.
      */
     def make: UIO[MemoMap] =
-      Ref.make[Map[ZLayer[Nothing, Any, Has[_]], Any]](Map.empty).map { ref =>
-        new MemoMap {
-          def get[E, A, B <: Has[_]](layer: ZLayer[A, E, B]): UIO[Option[B]] =
-            ref.get.map(_.get(layer).asInstanceOf[Option[B]])
-          def memoize[E, A, B <: Has[_]](layer: ZLayer[A, E, B], b: B): UIO[Unit] =
-            ref.update(_ + (layer -> b))
+      RefM
+        .make[Map[ZLayer[Nothing, Any, Has[_]], (Promise[_, _], Exit[Any, Any] => ZIO[Any, Nothing, Any], Ref[Int])]](
```

Comment:
A case class could be nicer for the value type.

Reply:
Good idea!